From bd4980987a954212e6bd122297a33dabf2281d37 Mon Sep 17 00:00:00 2001
From: wanghang73 <wanghang73@huawei.com>
Date: Sat, 6 Jun 2026 10:49:11 +0800
Subject: [PATCH] urma: improve urma_ping arg parsing and EID logging

(cherry picked from commit efc6324c38c7ca713bec39f48fa550ec17d46ccb)
---
 src/urma/tools/urma_ping/ping_parameters.c |  17 +-
 src/urma/tools/urma_ping/ping_parameters.h |   2 +-
 src/urma/tools/urma_ping/ping_run.c        | 172 +++++++++++----------
 3 files changed, 97 insertions(+), 94 deletions(-)

diff --git a/src/urma/tools/urma_ping/ping_parameters.c b/src/urma/tools/urma_ping/ping_parameters.c
index c68fd01..2f3930c 100644
--- a/src/urma/tools/urma_ping/ping_parameters.c
+++ b/src/urma/tools/urma_ping/ping_parameters.c
@@ -94,7 +94,7 @@ void usage(const char *filename)
               "  urma_ping [option] <destination>\n"
               "\n"
               "Options:\n"
-              "  <destination>    Primary eid or bonding eid\n"
+              "  <destination>    Primary EID or bonding EID\n"
               "  -c <count>       Packet count (default: INT_MAX)\n"
               "  -i <interval>    Interval between packets in seconds (default: 1)\n"
               "  -s <size>        Send buffer size in bytes (default: 4)\n"
@@ -116,7 +116,7 @@ int parse_args(ping_cfg_t *cfg)
     };
 
     int opt, ret = 0;
-    while ((opt = getopt_long(cfg->argc, cfg->argv, "c:i:s:qw:W:ShVv", long_options, NULL)) != -1) {
+    while ((opt = getopt_long(cfg->argc, cfg->argv, ":c:i:s:qw:W:ShVv", long_options, NULL)) != -1) {
         switch (opt) {
             case 'c':
                 ret = str_to_u32(optarg, &cfg->count);
@@ -146,29 +146,28 @@ int parse_args(ping_cfg_t *cfg)
                 version();
                 return -EINVAL;
             case ':':
-                LOG_ERROR("Option -%c requires an argument\n", optopt);
+                LOG_ERROR("Missing argument for option -%c\n", optopt);
                 return -EINVAL;
             default:
-                LOG_ERROR("Unknown option\n");
+                LOG_ERROR("Unknown option %s\n", cfg->argv[optind - 1]);
                 usage(cfg->filename);
                 return -EINVAL;
         }
         if (ret != 0) {
-            const int option_offset = 2;
-            LOG_ERROR("Invalid option %s\n", cfg->argv[optind - option_offset]);
+            LOG_ERROR("Invalid value for -%c: %s\n", opt, optarg);
             return -EINVAL;
         }
     }
 
     if (optind >= cfg->argc) {
-        LOG_ERROR("Destination is required\n");
+        LOG_ERROR("Destination EID is required\n");
         usage(cfg->filename);
         return -EINVAL;
     }
 
-    ret = str_to_eid(cfg->argv[optind], &cfg->dest);
+    ret = str_to_eid(cfg->argv[optind], &cfg->dst_eid);
     if (ret != 0) {
-        LOG_ERROR("Invalid dest eid %s\n", cfg->argv[optind]);
+        LOG_ERROR("Invalid destination EID: %s\n", cfg->argv[optind]);
         return -EINVAL;
     }
     return 0;
diff --git a/src/urma/tools/urma_ping/ping_parameters.h b/src/urma/tools/urma_ping/ping_parameters.h
index 12335e1..a3b5527 100644
--- a/src/urma/tools/urma_ping/ping_parameters.h
+++ b/src/urma/tools/urma_ping/ping_parameters.h
@@ -19,7 +19,7 @@ typedef struct ping_cfg {
     int argc;
     char **argv;
     char *filename;
-    urma_eid_t dest;
+    urma_eid_t dst_eid;
     uint32_t count;    /* UINT32_MAX means unlimited */
     uint32_t interval; /* seconds */
     uint32_t size;     /* bytes */
diff --git a/src/urma/tools/urma_ping/ping_run.c b/src/urma/tools/urma_ping/ping_run.c
index 1b3a98d..9e7c029 100644
--- a/src/urma/tools/urma_ping/ping_run.c
+++ b/src/urma/tools/urma_ping/ping_run.c
@@ -9,12 +9,12 @@
  */
 
 #include <malloc.h>
+#include <signal.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
 #include <unistd.h>
-#include <signal.h>
 
 #include "urma_api.h"
 #include "urma_types.h"
@@ -29,7 +29,6 @@
 const uint32_t PING_WK_JETTY_ID = 5;
 const uint32_t PING_SEND_DEPTH = 1024;
 const uint32_t PING_RECV_DEPTH = 1024;
-const uint32_t EID_STRLEN = 128;
 const uint32_t EID_MAX_VALUE = (1 << 30);
 
 typedef struct ping_urma_resource {
@@ -43,13 +42,6 @@ typedef struct ping_urma_resource {
     urma_target_jetty_t *tjetty;
 } ping_urma_resource_t;
 
-static void log_eid_verbose(uint8_t *eid)
-{
-    char print_str[EID_STRLEN];
-    inet_ntop(AF_INET6, eid, print_str, EID_STRLEN);
-    LOG_VERBOSE("%s\n", print_str);
-}
-
 static void primary_eid_to_main_primary_eid(urma_eid_t *primary_eid)
 {
     int ret;
@@ -94,13 +86,13 @@ static void primary_eid_to_main_primary_eid(urma_eid_t *primary_eid)
         for (int k = 0; k < IODIE_NUM; k++) {
             if (topo_map->topo_infos[target_node_id].agg_devs[j].ues[k].entity_id == target_entity_id) {
                 if (memcmp(topo_map->topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid,
-                    empty_eid.raw,
-                    EID_LEN) == 0) {
+                           empty_eid.raw,
+                           EID_LEN) == 0) {
                     continue;
                 }
                 if (memcmp(primary_eid->raw,
-                    topo_map->topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid,
-                    EID_LEN) > 0) {
+                           topo_map->topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid,
+                           EID_LEN) > 0) {
                     memcpy(primary_eid->raw,
                            topo_map->topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid,
                            EID_LEN);
@@ -135,7 +127,7 @@ static uint32_t get_chip_id_by_primary_eid(urma_eid_t *eid, urma_ping_ubcore_top
         }
     }
 
-    LOG_ERROR("Failed to get chip id by primary eid.\n");
+    LOG_ERROR("Failed to get chip ID for primary EID.\n");
     return 0;
 }
 
@@ -158,31 +150,24 @@ static bool get_primary_eid_and_chip_id_by_bonding(urma_eid_t *original_eid, urm
     if (is_primary_eid(original_eid, topo_map)) {
         memcpy(primary_eid->raw, original_eid->raw, EID_LEN);
         (*chip_id) = get_chip_id_by_primary_eid(original_eid, topo_map);
-        LOG_VERBOSE("Input dest eid is primary eid: ");
-        log_eid_verbose(primary_eid->raw);
-        LOG_VERBOSE("Chip id = %u.\n", (*chip_id));
         free(topo_map);
         return true;
     }
 
-    LOG_VERBOSE("Dest Bonding eid: ");
-    log_eid_verbose(original_eid->raw);
+    LOG_VERBOSE("Destination bonding EID: " EID_FMT "\n", EID_ARGS(*original_eid));
 
     for (int i = 0; i < (int)topo_map->node_num; i++) {
         for (int j = 0; j < DEV_NUM; j++) {
             if (memcmp(topo_map->topo_infos[i].agg_devs[j].agg_eid, original_eid->raw, EID_LEN) == 0) {
                 memcpy(primary_eid->raw, topo_map->topo_infos[i].agg_devs[j].ues[0].primary_eid, EID_LEN);
                 (*chip_id) = topo_map->topo_infos[i].agg_devs[j].ues[0].chip_id;
-                LOG_VERBOSE("Successful find dest primary eid: ");
-                log_eid_verbose(primary_eid->raw);
-                LOG_VERBOSE("Chip id = %u.\n", (*chip_id));
                 free(topo_map);
                 return true;
             }
         }
     }
 
-    LOG_ERROR("Failed to find dest primary eid.\n");
+    LOG_ERROR("Failed to find the destination primary EID.\n");
     free(topo_map);
     return false;
 }
@@ -203,8 +188,7 @@ static bool get_source_eid_by_bonding_eid_and_chip_id(urma_eid_t *eid, uint32_t
         return false;
     }
 
-    LOG_VERBOSE("Source Bonding eid: ");
-    log_eid_verbose(eid->raw);
+    LOG_VERBOSE("Source bonding EID: " EID_FMT "\n", EID_ARGS(*eid));
 
     for (int i = 0; i < (int)topo_map->node_num; i++) {
         for (int j = 0; j < DEV_NUM; j++) {
@@ -212,8 +196,7 @@ static bool get_source_eid_by_bonding_eid_and_chip_id(urma_eid_t *eid, uint32_t
                 for (int k = 0; k < IODIE_NUM; k++) {
                     if (topo_map->topo_infos[i].agg_devs[j].ues[k].chip_id == chip_id) {
                         memcpy(eid->raw, topo_map->topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN);
-                        LOG_VERBOSE("Successful find source primary eid: ");
-                        log_eid_verbose(eid->raw);
+                        LOG_VERBOSE("Source EID resolved to primary EID " EID_FMT "\n", EID_ARGS(*eid));
                         free(topo_map);
                         return true;
                     }
@@ -222,7 +205,7 @@ static bool get_source_eid_by_bonding_eid_and_chip_id(urma_eid_t *eid, uint32_t
         }
     }
 
-    LOG_ERROR("Failed to find source primary eid.\n");
+    LOG_ERROR("Failed to find the source primary EID.\n");
     free(topo_map);
     return false;
 }
@@ -238,7 +221,7 @@ static bool if_device_has_eid(urma_device_t *dev)
     return false;
 }
 
-static urma_device_t *find_first_urma_device()
+static urma_device_t *find_first_bonding_device()
 {
     int num_devices = 0;
     urma_device_t **devices = urma_get_device_list(&num_devices);
@@ -269,6 +252,48 @@ static void urma_log_func(int level, char *message)
     LOG_VVERBOSE("%s", message);
 }
 
+static int get_src_and_dst_eid(ping_cfg_t *cfg, urma_eid_t *src_eid, urma_eid_t *dst_eid)
+{
+    uint32_t chip_id = 0;
+    if (!get_primary_eid_and_chip_id_by_bonding(&(cfg->dst_eid), dst_eid, &chip_id)) {
+        return -ENODEV;
+    }
+
+    primary_eid_to_main_primary_eid(dst_eid);
+
+    LOG_VERBOSE("Destination EID resolved to primary EID " EID_FMT " (chip_id=%u)\n",
+                EID_ARGS(*dst_eid), chip_id);
+
+    urma_device_t *dev = find_first_bonding_device();
+    if (dev == NULL) {
+        LOG_ERROR("Failed to find first bonding device.\n");
+        return -ENODEV;
+    }
+
+    uint32_t eid_cnt = 1;
+    urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
+    if (eid_list == NULL) {
+        LOG_ERROR("Failed to get eid list.\n");
+        return -ENODEV;
+    }
+    if (eid_cnt != 1) {
+        LOG_ERROR("Multiple EIDs found for bonding_dev.\n");
+        urma_free_eid_list(eid_list);
+        return -ENODEV;
+    }
+
+    *src_eid = eid_list[0].eid;
+    urma_free_eid_list(eid_list);
+    eid_list = NULL;
+
+    if (!get_source_eid_by_bonding_eid_and_chip_id(src_eid, chip_id)) {
+        LOG_ERROR("Failed to resolve the source primary EID.\n");
+        return -ENODEV;
+    }
+
+    return 0;
+}
+
 static int init_urma_resource(ping_cfg_t *cfg, ping_urma_resource_t *res)
 {
     int ret = 0;
@@ -283,80 +308,57 @@ static int init_urma_resource(ping_cfg_t *cfg, ping_urma_resource_t *res)
         return ret;
     }
 
-    urma_device_t *dev = find_first_urma_device();
-    if (dev == NULL) {
-        LOG_ERROR("Failed to find first urma device.\n");
-        ret = -ENODEV;
+    urma_eid_t src_eid, dst_eid;
+    if ((ret = get_src_and_dst_eid(cfg, &src_eid, &dst_eid)) != 0) {
+        LOG_ERROR("Failed to get source and destination EIDs, ret:%d\n", ret);
         goto uninit_urma;
     }
 
-    uint32_t eid_cnt = 1, chip_id = 0, source_eid_index = EID_MAX_VALUE;
-    urma_eid_t source_eid, dest_eid;
-    if (!get_primary_eid_and_chip_id_by_bonding(&(cfg->dest), &dest_eid, &chip_id)) {
-        LOG_ERROR("Failed to dest primary eid.\n");
+    urma_device_t *dev = urma_get_device_by_eid(src_eid, URMA_TRANSPORT_UB);
+    if (dev == NULL) {
+        LOG_ERROR("Failed to find primary urma device.\n");
         ret = -ENODEV;
         goto uninit_urma;
     }
 
-    primary_eid_to_main_primary_eid(&dest_eid);
-
-    LOG_VERBOSE("Successful find main dest primary eid: ");
-    log_eid_verbose(dest_eid.raw);
-
-    urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
-    if (eid_list == NULL) {
-        LOG_ERROR("Failed to get eid list.\n");
-        ret = -ENODEV;
+    urma_device_attr_t dev_attr = {0};
+    if ((ret = urma_query_device(dev, &dev_attr)) != URMA_SUCCESS) {
+        LOG_ERROR("Failed to query urma device attr, ret:%d\n", ret);
         goto uninit_urma;
     }
-    if (eid_cnt != 1) {
-        LOG_ERROR("More than one eid for bonding_dev.\n");
-        ret = -ENODEV;
+    if (cfg->size > dev_attr.dev_cap.max_msg_size) {
+        LOG_ERROR("Ping message size %u exceeds device max msg size %u.\n",
+                  cfg->size, dev_attr.dev_cap.max_msg_size);
+        ret = -EINVAL;
         goto uninit_urma;
     }
 
-    source_eid = eid_list[0].eid;
-    if (!get_source_eid_by_bonding_eid_and_chip_id(&source_eid, chip_id)) {
-        LOG_ERROR("Failed to source primary eid.\n");
-        ret = -ENODEV;
-        goto uninit_urma;
-    }
+    LOG_VERBOSE("Selected device: %s\n", dev->name);
 
-    urma_device_t *primary_dev = urma_get_device_by_eid(source_eid, URMA_TRANSPORT_UB);
-    if (primary_dev == NULL) {
-        LOG_ERROR("Failed to find primary urma device.\n");
+    uint32_t src_eid_idx = EID_MAX_VALUE;
+    uint32_t eid_cnt;
+    urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
+    if (eid_list == NULL) {
+        LOG_ERROR("Failed to get source device EID list.\n");
         ret = -ENODEV;
         goto uninit_urma;
     }
-    LOG_VERBOSE("device = %s\n", primary_dev->name);
-
-    eid_list = urma_get_eid_list(primary_dev, &eid_cnt);
     for (int i = 0; i < eid_cnt; i++) {
-        if (memcmp(&source_eid, &(eid_list[i].eid), sizeof(urma_eid_t)) == 0) {
-            source_eid_index = eid_list[i].eid_index;
+        if (memcmp(&src_eid, &(eid_list[i].eid), sizeof(urma_eid_t)) == 0) {
+            src_eid_idx = eid_list[i].eid_index;
             break;
         }
     }
-    if (source_eid_index == EID_MAX_VALUE) {
-        LOG_ERROR("Failed to find primary eid_index.\n");
+    urma_free_eid_list(eid_list);
+    if (src_eid_idx == EID_MAX_VALUE) {
+        LOG_ERROR("Failed to find the source EID index.\n");
         ret = -ENODEV;
         goto uninit_urma;
     }
-    LOG_VERBOSE("Get primary eid_index = %d\n", (int)source_eid_index);
 
-    urma_device_attr_t dev_attr = {0};
-    if ((ret = urma_query_device(primary_dev, &dev_attr)) != URMA_SUCCESS) {
-        LOG_ERROR("Failed to query urma device attr, ret:%d\n", ret);
-        goto uninit_urma;
-    }
-    if (cfg->size > dev_attr.dev_cap.max_msg_size) {
-        LOG_ERROR("Ping message size %u exceeds device max msg size %u.\n",
-                  cfg->size, dev_attr.dev_cap.max_msg_size);
-        ret = -EINVAL;
-        goto uninit_urma;
-    }
+    LOG_VERBOSE("Source EID index: %d\n", src_eid_idx);
 
-    res->ctx = urma_create_context(primary_dev, source_eid_index);
+    res->ctx = urma_create_context(dev, src_eid_idx);
     if (res->ctx == NULL) {
         LOG_ERROR("Failed to create urma context.\n");
         ret = -EINVAL;
@@ -436,7 +438,7 @@ static int init_urma_resource(ping_cfg_t *cfg, ping_urma_resource_t *res)
     }
 
     urma_rjetty_t rjetty = {
-        .jetty_id.eid = dest_eid,
+        .jetty_id.eid = dst_eid,
         .jetty_id.id = PING_WK_JETTY_ID,
         .trans_mode = URMA_TM_RM,
         .type = URMA_JETTY,
@@ -477,6 +479,8 @@ static void uninit_urma_resource(ping_urma_resource_t *res)
     urma_unimport_jetty(res->tjetty);
     urma_delete_jetty(res->jetty);
     urma_delete_jfr(res->jfr);
+    urma_delete_jfc(res->recv_jfc);
+    urma_delete_jfc(res->send_jfc);
     urma_unregister_seg(res->seg);
     free(res->buf);
     res->buf = NULL;
@@ -578,11 +582,11 @@ static int recv_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_se
             // success
             update_stat_on_recv(seq_info->rtt);
             LOG_NORMAL("%d bytes from " EID_FMT ": seq=%d time=%.3f ms\n",
-                       cfg->size, EID_ARGS(cfg->dest), seq_info->seq, seq_info->rtt);
+                       cfg->size, EID_ARGS(cfg->dst_eid), seq_info->seq, seq_info->rtt);
         } else {
             // failure
             LOG_NORMAL("From " EID_FMT " seq=%d cr_status=%d\n",
-                       EID_ARGS(cfg->dest), seq_info->seq, cr.status);
+                       EID_ARGS(cfg->dst_eid), seq_info->seq, cr.status);
         }
     }
 
@@ -591,7 +595,7 @@ static int recv_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_se
 
 static void signal_handler(int signum)
 {
-    LOG_VERBOSE("Received Signal: %d\n", signum);
+    LOG_VERBOSE("Received signal %d\n", signum);
     print_stat();
     exit(0);
 }
@@ -612,7 +616,7 @@ int start_ping(ping_cfg_t *cfg)
         return ret;
     }
 
-    LOG_QUIET("URMA_PING " EID_FMT " %u bytes of data.\n", EID_ARGS(cfg->dest), cfg->size);
+    LOG_QUIET("URMA_PING " EID_FMT " %u bytes of data.\n", EID_ARGS(cfg->dst_eid), cfg->size);
     LOG_VERBOSE("Count       : %u\n", cfg->count);
     LOG_VERBOSE("Interval(s) : %u\n", cfg->interval);
     LOG_VERBOSE("Size(bytes) : %u\n", cfg->size);
@@ -649,7 +653,7 @@ int start_ping(ping_cfg_t *cfg)
 
         LOG_VERBOSE("Time elapsed = %.3lfms\n", time_elapsed);
         if (cfg->deadline != 0 && time_elapsed > cfg->deadline * 1000.0) {
-            LOG_ERROR("URMA_PING time limit exceeded.\n");
+            LOG_NORMAL("Deadline reached after %.3f ms, stopping.\n", time_elapsed);
             break;
         }
     }
-- 
2.34.1

